### DISCUSSION TITLE
initial bootstrap w/ bitbucket server

### DISCUSSION THREAD

**Discussion by mdnghtsun** [fluxcd/flux2#3036] [Q&A]
I'm new to fluxcd and I'm having trouble running the initial boot strap command with bitbucket server:

> ◎ waiting for Kustomization "flux-system/flux-system" to be reconciled
> ✗ client rate limiter Wait returned an error: context deadline exceeded
> ...
> ✗ bootstrap failed with 1 health check failure(s)

### flux bootstrap
> $ flux bootstrap bitbucket-server --verbose   --token-auth   --hostname=bitbucket.venom.com   --owner=blah   --username=sunny.mundra   --repository=apollo-fluxcd   --branch=master   --path=clusters/test-cluster --insecure-skip-tls-verify
> ► connecting to bitbucket.venom.com
> ► cloning branch "master" from Git repository "https://bitbucket.venom.com/scm/blah/apollo-fluxcd.git"
> ✔ cloned repository
> ► generating component manifests
> ✔ generated component manifests
> ✔ component manifests are up to date
> ► installing components in "flux-system" namespace
> ✔ installed components
> ✔ reconciled components
> ► determining if source secret "flux-system/flux-system" exists
> ► generating source secret
> ► applying source secret "flux-system/flux-system"
> ✔ reconciled source secret
> ► generating sync manifests
> ✔ generated sync manifests
> ✔ sync manifests are up to date
> ► applying sync manifests
> ✔ reconciled sync configuration
> ◎ waiting for Kustomization "flux-system/flux-system" to be reconciled
> ✗ client rate limiter Wait returned an error: context deadline exceeded
> ► confirming components are healthy
> ✔ helm-controller: deployment ready
> ✔ kustomize-controller: deployment ready
> ✔ notification-controller: deployment ready
> ✔ source-controller: deployment ready
> ✔ all components are healthy
> ✗ bootstrap failed with 1 health check failure(s)

Can you help me understand whats going on here? Some more information:

### kubectl get pod
> $ kubectl get pod -n flux-system
> NAME                                       READY   STATUS    RESTARTS   AGE
> helm-controller-97cb7b49f-bn6fd            1/1     Running   0          4m59s
> kustomize-controller-684ccd454b-lsxmm      1/1     Running   0          4m59s
> notification-controller-79c746b8c4-dzstd   1/1     Running   0          4m59s
> source-controller-69dd4bb77c-mstht         1/1     Running   0          4m59s

### kubectl describe pod
> Events:
>   Type    Reason     Age    From               Message
>   ----    ------     ----   ----               -------
>   Normal  Scheduled  7m18s  default-scheduler  Successfully assigned flux-system/helm-controller-97cb7b49f-bn6fd to ip-192-168-18-246.ec2.internal
>   Normal  Pulled     7m17s  kubelet            Container image "ghcr.io/fluxcd/helm-controller:v0.22.2" already present on machine
>   Normal  Created    7m17s  kubelet            Created container manager
>   Normal  Started    7m17s  kubelet            Started container manager

### flux log
> $ flux logs
> 2022-08-25T22:26:17.479Z info  - Metrics server is starting to listen
> 2022-08-25T22:26:17.480Z info  - starting manager
> 2022-08-25T22:26:17.481Z info  - Starting server
> 2022-08-25T22:26:17.481Z info  - Starting server
> 2022-08-25T22:26:17.602Z info HelmRelease - Starting EventSource
> 2022-08-25T22:26:17.602Z info HelmRelease - Starting EventSource
> 2022-08-25T22:26:17.602Z info HelmRelease - Starting Controller
> 2022-08-25T22:26:17.703Z info HelmRelease - Starting workers
> 2022-08-25T22:26:17.861Z info  - Metrics server is starting to listen
> 2022-08-25T22:26:17.862Z info  - starting manager
> 2022-08-25T22:26:17.862Z info  - Starting server
> 2022-08-25T22:26:17.862Z info  - Starting server
> 2022-08-25T22:26:17.980Z info Kustomization - Starting EventSource
> 2022-08-25T22:26:17.980Z info Kustomization - Starting EventSource
> 2022-08-25T22:26:17.980Z info Kustomization - Starting EventSource
> 2022-08-25T22:26:17.980Z info Kustomization - Starting Controller
> 2022-08-25T22:26:18.081Z info Kustomization - Starting workers
> 2022-08-25T22:26:17.810Z info  - Metrics server is starting to listen
> 2022-08-25T22:26:17.812Z info  - starting event server
> 2022-08-25T22:26:17.812Z info  - starting webhook receiver server
> 2022-08-25T22:26:17.812Z info  - starting manager
> 2022-08-25T22:26:17.812Z info  - Starting server
> 2022-08-25T22:26:17.812Z info  - Starting server
> 2022-08-25T22:26:17.934Z info Provider - Starting EventSource
> 2022-08-25T22:26:17.934Z info Provider - Starting Controller
> 2022-08-25T22:26:17.934Z info Receiver - Starting EventSource
> 2022-08-25T22:26:17.934Z info Receiver - Starting Controller
> 2022-08-25T22:26:17.934Z info Alert - Starting EventSource
> 2022-08-25T22:26:17.934Z info Alert - Starting EventSource
> 2022-08-25T22:26:17.934Z info Alert - Starting Controller
> 2022-08-25T22:26:18.035Z info Receiver - Starting workers
> 2022-08-25T22:26:18.035Z info Provider - Starting workers
> 2022-08-25T22:26:18.035Z info Alert - Starting workers
> 2022-08-25T22:26:18.554Z info  - Metrics server is starting to listen
> 2022-08-25T22:26:18.555Z info  - starting manager
> 2022-08-25T22:26:18.556Z info  - Starting server
> 2022-08-25T22:26:18.556Z info  - Starting server
> 2022-08-25T22:26:18.674Z info GitRepository - Starting EventSource
> 2022-08-25T22:26:18.674Z info GitRepository - Starting Controller
> 2022-08-25T22:26:18.674Z info HelmRepository - Starting EventSource
> 2022-08-25T22:26:18.674Z info HelmRepository - Starting Controller
> 2022-08-25T22:26:18.675Z info HelmRepository - Starting EventSource
> 2022-08-25T22:26:18.675Z info HelmRepository - Starting Controller
> 2022-08-25T22:26:18.675Z info Bucket - Starting EventSource
> 2022-08-25T22:26:18.675Z info Bucket - Starting Controller
> 2022-08-25T22:26:18.675Z info  - starting file server
> 2022-08-25T22:26:18.675Z info HelmChart - Starting EventSource
> 2022-08-25T22:26:18.675Z info HelmChart - Starting EventSource
> 2022-08-25T22:26:18.675Z info HelmChart - Starting EventSource
> 2022-08-25T22:26:18.676Z info HelmChart - Starting EventSource
> 2022-08-25T22:26:18.676Z info HelmChart - Starting Controller
> 2022-08-25T22:26:18.775Z info HelmRepository - Starting workers
> 2022-08-25T22:26:18.775Z info GitRepository - Starting workers
> 2022-08-25T22:26:18.775Z info Bucket - Starting workers
> 2022-08-25T22:26:18.776Z info HelmRepository - Starting workers
> 2022-08-25T22:26:18.776Z info HelmChart - Starting workers

**Comment by kingdonb** [comment]
We are talking about this at Flux Bug Scrub today

> `bootstrap failed with 1 health check failure(s)`

This could be a transient error that the pods are taking a bit of extra time to get started, if you don't find any problems and Flux seems to be working this is most likely the case.

We think there is some room for improvement here, the output could tell you what the health check failures represent 

> `✗ client rate limiter Wait returned an error: context deadline exceeded`

This could be due to your internet connection, or some other hard to quantify source of errors. Is it consistent every time? (If you bootstrap again with an empty target path, and a new cluster, do you always get the same error? If you run flux bootstrap a second time on the same cluster with the same commands, does it succeed the second time?)

The version of Flux that you have installed is not very old but also not the newest. Can you try on a more recent version of the Flux CLI as well?

**Reply by mdnghtsun** [reply]
@kingdonb Yes, its consistent every time. 

Running bootstrap without a target path gives a different error:
`[ec2-user@ip-10-0-10-228 ~]$ flux bootstrap bitbucket-server --verbose   --token-auth   --hostname=bitbucket.venom.com   --owner=blah   --username=sunny.mundra   --repository=apollo-fluxcd   --branch=master    --insecure-skip-tls-verify
► connecting to bitbucket.venom.com
► cloning branch "master" from Git repository "https://bitbucket.venom.com/scm/blah/apollo-fluxcd.git"
✔ cloned repository
► generating component manifests
✔ generated component manifests
✔ committed sync manifests to "master" ("40f06e24750057abc6d6481e1efb41aec7240de3")
► pushing component manifests to "https://bitbucket.venom.com/scm/blah/minerva-fluxcd.git"
► installing components in "flux-system" namespace
✔ installed components
✔ reconciled components
► determining if source secret "flux-system/flux-system" exists
► generating source secret
► applying source secret "flux-system/flux-system"
✔ reconciled source secret
✗ sync path configuration ("") would overwrite path ("./clusters/test-cluster") of existing Kustomization`

Running bootstrap with the same command again gives the same error as shared originally.

**Reply by kingdonb** [reply]
> If you bootstrap again with an empty target path, and a new cluster, do you always get the same error?

I'm not sure what I was going for here, or why you should use an empty target path... sorry, I must have been distracted when I wrote this. Oh! By "empty target path" I meant to set `--path=./clusters/new-test-cluster` rather than `--path=./clusters/test-cluster` or the implicit `--path=''` empty string path. My fault, I wasn't clear.

> `✗ client rate limiter Wait returned an error: context deadline exceeded`

I'm not sure what could cause this error except for some kind of network issue between your workstation and the cluster. We have definitely seen this type of thing before, but I unfortunately can't narrow it down any further without some hands-on debugging.

If you run `flux check` after the installation, do you see any issues there? (Can you paste the output here?) Also, Protip-If you use three backticks on a line by itself to start and end the code block in your comment, like this (without the slashes):
```
\```
codeblock
\```
```
then you will get better markdown formatting, (it's harder to read when everything runs together on one line.)
